src/node_realm.cc

```diff
@@ -269,8 +282,67 @@ void Realm::DoneBootstrapping() {
   CHECK(env_->req_wrap_queue()->IsEmpty());
   CHECK(env_->handle_wrap_queue()->IsEmpty());
 
-  // TODO(legendecas): track base object count by realms.
-  env_->set_base_object_created_by_bootstrap(env_->base_object_count());
+  has_run_bootstrapping_code_ = true;
+
+  // This adjusts the return value of base_object_created_after_bootstrap() so
+  // that tests that check the count do not have to account for internally
+  // created BaseObjects.
+  base_object_created_by_bootstrap_ = base_object_count_;
+}
+
+void Realm::RunCleanup() {
+  TRACE_EVENT0(TRACING_CATEGORY_NODE1(realm), "RunCleanup");
+
+  cleanup_queue_.Drain();
+}
+
+void Realm::PrintInfoForSnapshotIfDebug() {
+  fprintf(stderr, "Realm:\n");
+  if (env()->enabled_debug_list()->enabled(DebugCategory::MKSNAPSHOT)) {
+    fprintf(stderr, "BaseObjects of the Realm:\n");
+    PrintAllBaseObjects();
+  }
+  fprintf(stderr, "End of the Realm.\n");
+}
+
+void Realm::PrintAllBaseObjects() {
+  size_t i = 0;
+  std::cout << "BaseObjects\n";
+  ForEachBaseObject([&](BaseObject* obj) {
+    std::cout << "#" << i++ << " " << obj << ": " << obj->MemoryInfoName()
+              << "\n";
+  });
+}
+
+void Realm::VerifyNoStrongBaseObjects() {
+  // When a process exits cleanly, i.e. because the event loop ends up without
+  // things to wait for, the Node.js objects that are left on the heap should
+  // be:
+  //
+  //   1. weak, i.e. ready for garbage collection once no longer referenced, or
+  //   2. detached, i.e. scheduled for destruction once no longer referenced, or
+  //   3. an unrefed libuv handle, i.e. does not keep the event loop alive, or
+  //   4. an inactive libuv handle (essentially the same here)
+  //
+  // There are a few exceptions to this rule, but generally, if there are
+  // C++-backed Node.js objects on the heap that do not fall into the above
+  // categories, we may be looking at a potential memory leak. Most likely,
+  // the cause is a missing MakeWeak() call on the corresponding object.
+  //
+  // In order to avoid this kind of problem, we check the list of BaseObjects
+  // for these criteria. Currently, we only do so when explicitly instructed to
+  // or when in debug mode (where --verify-base-objects is always-on).
+
+  if (!env()->options()->verify_base_objects) return;
```

Comment:
I wonder if it's necessary to introduce per-realm options (maybe not in this PR, though)

Reply:
I believe it is a good idea to distinguish per-realm options. I've added a TODO comment here. Thank you for the suggestion.